### ISSUE TITLE
Integrate vi mode with clipboard

### ISSUE THREAD

**Issue by jamesla** [fish-shell/fish-shell#4028]
I'm trying to get fish vi mode to use the clipboard clipboard meaning y and p will yank and paste from the clipboard that the rest of the system is using.

I have had a look through the docs and existing issues but was unable to figure out how to do it. How can I achieve this with fish?

**Comment by faho** [comment]
So, we used to sync our clipboard with the X11 one, but we removed it because it was confusing and annoying (e.g. when you copy something from another application and then clear the line so you can paste it it would have replaced the clipboard contents with fish's previous line).

Instead, we have bindings to explicitly copy/paste. These are bound to ctrl-x for copy (since ctrl-c is cancel) and ctrl-v for paste. Currently, they can only copy the entire line.

You can bind "yy" and "Y" to copy the entire line easily:

```fish
bind yy fish_clipboard_copy
bind Y fish_clipboard_copy
bind p fish_clipboard_paste
```

The problem is getting the other bindings ("yiw" and such) working. One possible approach is to make fish's "killring" accessible to scripts, so you could make a function that syncs it with the system clipboard. That would then look like

```fish
bind yW 'commandline -f kill-bigword yank; fish_clipboard_copy $killring[1]'
```

**Comment by jamesla** [comment]
Thanks Faho, when running the command fish_clipboard_paste from shell it is pasting the correct text however  when I added the following to my config.fish

```
function fish_user_key_bindings
  bind yy fish_clipboard_copy
  bind Y fish_clipboard_copy
  bind p fish_clipboard_paste
end

```

It still seems to be using the primary clipboard. Is there anything else you can think of?

**Comment by faho** [comment]
>It still seems to be using the primary clipboard. Is there anything else you can think of?

Humm... When you say "primary clipboard", are you referring to the X11 "primary selection" (i.e. the middle-click paste thing)?

I've just added this to my config, then I entered something, switched to normal mode, pressed "yy" and pasted it (via the terminal's ctrl-shift-v binding - which executes outside of fish). The text I just pasted appeared. I then copied text in firefox and pressed "p" in fish, and that text appeared. So it seems to be working for me.

**Comment by Ambrevar** [comment]
Is there a way to copy/paste to/from the primary selection?

**Comment by faho** [comment]
@Ambrevar: Not specifically. You'll need to either modify or copy the `fish_clipboard_*` functions for that - change "--clipboard" to "--primary".

**Comment by Ambrevar** [comment]
Plans for adding `fish_primary_paste` and `fish_primary_copy`?

**Comment by faho** [comment]
No. Personally, I don't use it. If anyone wants it, it's quite trivial to change it yourself.

**Comment by jamesla** [comment]
@faho Sorry new to fish shell, how would you change it yourself?

**Comment by krader1961** [comment]
@jamesla: Run `funced fish_clipboard_copy`. Replace `--clipboard` with `--primary` and exit your editor. Then run `funcsave fish_clipboard_copy` to write it to your personal functions directory. Do the same thing for `fish_clipboard_paste`.

**Comment by jamesla** [comment]
Ok so here is what I have done.

1.  `funced fish_clipboard_copy`


2. `function fish_clipboard_copy
                             if type -q pbcopy
                                 commandline | pbcopy
                             else if type -q xsel
                                 commandline | xsel --primary
                             end
                         end
` 

3. enter insert mode and type test123
4. yy to copy line
5. p to test it pastes correctly (it does)
6. `xclip -o -selection XA_CLIPBOARD`
7. `xclip -o -selection XA_PRIMARY`

Neither of the xclip outputs contain test123.

I feel like I must be missing something here. Any ideas?

**Comment by faho** [comment]
>I feel like I must be missing something here. Any ideas?

@jamesla: First, the obvious: Do you actually have `xsel` installed? If not, `fish_clipboard_copy` won't do anything.

I'd be willing to take a patch to add an `xclip` path to it (and `fish_clipboard_paste`).

**Comment by jamesla** [comment]
Sorry I should have used xsel for the example yes it is installed and functioning.

1.  `funced fish_clipboard_copy`
2. `function fish_clipboard_copy
                             if type -q pbcopy
                                 commandline | pbcopy
                             else if type -q xsel
                                 commandline | xsel --primary
                             end
                         end
` 
3. enter insert mode and type test123
4. yy to copy line
5. p to test it pastes correctly (it does)
6. `xsel -o ---clipboard`
7. `xsel -o --primary`

Neither of the xclip outputs contain test123.

Second test

1. `echo test123 | xsel -i --primary`
2. `xsel -o --primary` 

This works so no problem with my clipboard.

Am I using this in line with the initial intention when the feature was created? If so is this working for anybody else?

**Comment by faho** [comment]
Aaaah, sorry.... the vi bindings don't use fish_clipboard_copy in this case.

They actually still use fish's internal killring.

TBD is what we should do about it. Ideally we'd expose the killring via a variable, and then just run `fish_clipboard_copy $killring[1]` in the bindings.

**Comment by faho** [comment]
As a workaround, use ctrl+x to copy the entire line - that's defined in the shared bindings, so it's also available in vi-mode.

**Comment by jamesla** [comment]
Ok thanks for clearing that up, I'll close the issue.

**Comment by Ossur** [comment]
In vim the `+` register functions also like an "output pipe" to the clipboard, so you can copy to the clipboard with `"+y` or cut to the clipboard with `"+d`. Would it be possible to make the `+` register behave in the same special way in fish vi mode?

**Comment by faho** [comment]
>Would it be possible to make the + register behave in the same special way in fish vi mode?

Possible? Sure. But it's going to need a whole bunch of work unless you really just want `"*p` to paste from the system clipboard (which it currently actually does, except it doesn't use `fish_clipboard_paste`) and `"*y` to copy to it.

Because currently our vi mode has no concept of an "argument" to a "command". So you can use "dw" to delete a word because that's explicitly bound, but you can't use "d2w" to delete two words because "d2" isn't. And binding all possible combinations is unfeasible, so that concept would have to be introduced.

Though now that I think about it, binding those two commands would probably solve 90% of the issue.

**Comment by tamis-laan** [comment]
Would love this 

> So, we used to sync our clipboard with the X11 one, but we removed it because it was confusing and annoying (e.g. when you copy something from another application and then clear the line so you can paste it it would have replaced the clipboard contents with fish's previous line).
> 
> Instead, we have bindings to explicitly copy/paste. These are bound to ctrl-x for copy (since ctrl-c is cancel) and ctrl-v for paste. Currently, they can only copy the entire line.
> 
> You can bind "yy" and "Y" to copy the entire line easily:
> 
> ```fish
> bind yy fish_clipboard_copy
> bind Y fish_clipboard_copy
> bind p fish_clipboard_paste
> ```
> 
> The problem is getting the other bindings ("yiw" and such) working. One possible approach is to make fish's "killring" accessible to scripts, so you could make a function that syncs it with the system clipboard. That would then look like
> 
> ```fish
> bind yW 'commandline -f kill-bigword yank; fish_clipboard_copy $killring[1]'
> ```

Would love to have fish shell copy/paste work like vim properly!

**Comment by melissaboiko** [comment]
I would like this for command-line editing in general. Being an emacs user, I kinda expect that if I C-k (kill-line) something, that something will be in the X clipboard for other apps to use.

**Comment by faho** [comment]
@melissaboiko Fish used to have that, by calling `xsel` constantly. This was super surprising to most users, it's not something we wish to revisit. See #3061.

**Comment by melissaboiko** [comment]
@faho Is there way to redefine the keybound functions so that I can do the delete-line and also copy it to clipboard using xsel?

**Comment by faho** [comment]
No. The issue is that the key binding functions will be *queued* up, so you can't do one and react to what it changed in script.

**Comment by tamis-laan** [comment]
The binding `bind p fish_clipboard_paste` now produces the wrong result. 

When you paste nothing happends unltil you move the cursor, then the character is replaced by the thing being pasted:

`wor**d**` --paste--> `word` --move cursor--> `worbla`

**Comment by yanshay** [comment]
> > ```fish
> > bind yy fish_clipboard_copy
> > bind Y fish_clipboard_copy
> > bind p fish_clipboard_paste
> > ```
> 
> thanks, it works for Y in normal mode
> 
> how can i make y in visual mode copying selected text?

I have the same question. How can I make 'y' in vi visial mode copy selected text to the clipboard?

**Comment by tamis-laan** [comment]
``` fish
bind yy fish_clipboard_copy
bind Y fish_clipboard_copy
bind p fish_clipboard_paste
```

This will paste in place. The other thing I'm missing is `cc` 'replace line' and as @yanshay says yanking in visual mode.

**Comment by primeapple** [comment]
Also I'm wondering how to "overpaste" selected text with the clipboard

**Comment by CWEzio** [comment]
I would like to make "+y yank to clipboard as in vim. (Currently, it is "*y, which in vim copy to the middle click selection.) I try to use the following key to implement it
```
     bind --mode visual \"+y fish_clipboard_copy end-selection --sets-mode default repaint-mode
```
However, it does as expected. Though it does do the copy work, it does not cancel the visual selection. Even the mode has went back to default. Any idea how to solve this problem?

**Comment by timhansinger** [comment]
> > bind yy fish_clipboard_copy
> > bind Y fish_clipboard_copy
> > bind p fish_clipboard_paste
> 
> thanks, it works for Y in normal mode
> 
> how can i make y in visual mode copying selected text?

I now added this into my function fish_user_key_bindings. With this I can use "Y" to copy the in visual mode selected text to my system clipboard:

``` fish
bind -s --preset -M visual -m default Y "fish_clipboard_copy; commandline -f end-selection repaint-mode"
```

**Comment by desgua** [comment]
The problem with the bind:
```
  bind p fish_clipboard_paste
```

is that the clipboard content is pasted BEFORE the cursor position, which is annoying for those who are used with the vi convention to paste AFTER the cursor position.
One can fix it by modifying the file:

> /usr/share/fish/functions/__fish_paste.fish

  change the line:
```
    commandline -i -- $data
```

  to
```
   commandline -a -- $data
   commandline -f end-of-line
```

**Comment by chamabreu** [comment]
> The problem with the bind:
> 
> ```
>   bind p fish_clipboard_paste
> ```
> 
> is that the clipboard content is pasted BEFORE the cursor position, which is annoying for those who are used with the vi convention to paste AFTER the cursor position. One can fix it by modifying the file:
> 
> > /usr/share/fish/functions/__fish_paste.fish
> 
> change the line:
> 
> ```
>     commandline -i -- $data
> ```
> 
> to
> 
> ```
>    commandline -a -- $data
>    commandline -f end-of-line
> ```

Nice, I did not edit the system file, but just used the command directly:
`bind -M default p 'commandline -f forward-single-char; fish_clipboard_paste'`

**Comment by krobelus** [comment]
> The problem with the bind:
>   bind p fish_clipboard_paste
>
> is that the clipboard content is pasted BEFORE the cursor position

no, that has been fixed in recent versions

**Comment by primeapple** [comment]
> Also I'm wondering how to "overpaste" selected text with the clipboard

```
bind --mode visual --sets-mode default p "commandline -f kill-selection end-selection; fish_clipboard_paste; set fish_bind_mode default; commandline -f repaint-mode"
```

did it for me.
